src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -189,40 +207,129 @@ irep_idt generate_java_generic_typet::build_generic_tag(
-      // arrays with different types.
-      if(element_type.id() == ID_pointer)
+      PRECONDITION(is_java_generic_parameter(parameter));
+      const std::string &parameter_name = to_java_generic_parameter(parameter)
+                                            .type_variable()
+                                            .get_identifier()
+                                            .c_str();
+      // get the class name without the `java::` prefix, e.g., for
+      // `java::Class::T` parameter_class_name is `Class`
+      const std::string &parameter_class_name =
+        parameter_name.substr(6, parameter_name.rfind("::") - 6);
+
+      if(parameter_class_name != current_outer_class_name)
       {
-        const symbol_typet element_symbol =
-          to_symbol_type(element_type.subtype());
-        new_tag_buffer << "of_" << id2string(element_symbol.get_identifier());
+        // we move on to the next class, which is an inner class of the
+        // current_outer_class_name, i.e., current_outer_class_name is a
+        // prefix of parameter_class_name
+        PRECONDITION(
+          parameter_class_name.compare(current_outer_class_name) > 0);
+        if(!first_outer)
+          new_tag_buffer << ">";
+        first_outer = false;
+        new_tag_buffer << parameter_class_name.substr(
+          current_outer_class_name.length(), std::string::npos);
+        current_outer_class_name = parameter_class_name;
+        new_tag_buffer << "<";
+        first = true;
+      }
+
+      if(!first)
+        new_tag_buffer << ", ";
+      first = false;
+
+      const irep_idt &id(
+        id2string(generic_argument_p->subtype().get(ID_identifier)));
+      new_tag_buffer << pretty_print_java_type(id2string(id));
+      if(is_java_array_tag(id))
+      {
+        const typet &element_type = java_array_element_type(
+          to_symbol_type(generic_argument_p->subtype()));
+
+        // If this is an array of references then we will specialize its
+        // identifier using the type of the objects in the array. Else, there
+        // can be a problem with the same symbols for different instantiations
+        // using arrays with different types.
+        if(element_type.id() == ID_pointer)
+        {
+          const symbol_typet element_symbol =
+            to_symbol_type(element_type.subtype());
+          new_tag_buffer << "of_" << id2string(element_symbol.get_identifier());
+        }
       }
+      ++generic_argument_p;
     }
+
+    new_tag_buffer << ">";
+    new_tag_buffer << original_class_tag.substr(
+      current_outer_class_name.length(), std::string::npos);
+  }
+  else
+  {
+    new_tag_buffer << original_class_tag;
   }
 
-  new_tag_buffer << ">";
+  // if the original class is generic, add type arguments for the class itself
+  if(is_java_generic_class_type(original_class))
+  {
+    new_tag_buffer << "<";
+    bool first = true;
+    while(generic_argument_p !=
```

Comment:
Is there a reason to use a `while` loop here rather than a for loop as you've used above